implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex

```diff
@@ -1,380 +1,611 @@
+    {:reply, {:error, :handshake_not_finished}, state}
+  end
 
-    metadata = Map.put(metadata, :return_value, return_value)
-    Telemetry.emit_stop_event([__MODULE__, :handle_event], start_time, metadata: metadata)
+  def handle_call(
+        :disconnect,
+        _from,
+        %{state: %Channel{channel_state: %Established{} = e} = s} = ws
+      ) do
+    payload = ServiceMessage.encode!(%ServiceMessage{command: :disconnect})
+    msg = %Message{onward_route: [], return_route: [], payload: payload}
+    send_over_encrypted_channel(msg, e.encrypt_st, s.peer_route, s.inner_address)
+    {:stop, :normal, :ok, ws}
+  end
 
-    return_value
+  def handle_call(:established?, _from, %{state: %Channel{channel_state: %Established{}}} = ws) do
+    {:reply, true, ws}
   end
 
-  defp do_handle_event({:call, from}, :established?, state, data) do
-    established = {:encrypted_transport, :ready} === state
-    {:next_state, state, data, [{:reply, from, established}]}
+  def handle_call(:established?, _from, ws) do
+    {:reply, false, ws}
   end
 
-  defp do_handle_event(:internal, event, {:key_establishment, _role, _role_state} = state, data) do
-    key_establishment_protocol = Map.get(data, :key_establishment_protocol)
-    key_establishment_protocol.handle_internal(event, state, data)
+  def handle_call(:role, _from, %{state: %Channel{role: role}} = ws) do
+    {:reply, role, ws}
   end
 
-  defp do_handle_event(:info, %Ockam.Message{} = message, state, data) do
-    ## TODO: make secure channel a worker
-    result =
-      Ockam.Worker.with_handle_message_metric(__MODULE__, message, data, fn ->
-        case is_authorized(message, data) do
-          :ok ->
-            ## reply is using gen_statem format,
-            ## not supported by :with_handle_message_metric
-            case handle_message(message, state, data) do
-              {:error, reason} -> {:error, reason}
-              {:stop, reason, data} -> {:stop, reason, data}
-              other -> {:ok, other}
-            end
+  defp worker_return({:ok, channel_state}, worker_state),
+    do: {:ok, Map.put(worker_state, :state, channel_state)}
 
-          {:error, reason} ->
-            {:error, {:unauthorized, reason}}
-        end
-      end)
+  defp worker_return({:error, reason}, worker_state), do: {:stop, {:error, reason}, worker_state}
 
-    ## handle_message_result is using gen_statem format,
-    ## not supported by :with_handle_message_metric
-    case result do
-      {:ok, handle_message_result} -> handle_message_result
-      other -> other
-    end
-  end
+  defp worker_return({:stop, reason, channel_state}, worker_state),
+    do: {:stop, reason, Map.put(worker_state, :state, channel_state)}
+
+  defp new_xx(vault, opts) do
+    m1_payload = m2_payload = m3_payload = ""
+    {:ok, e_key_handle} = Ockam.Vault.secret_generate(vault, type: :curve25519)
+    {:ok, e_public_key} = Vault.secret_publickey_get(vault, e_key_handle)
```

Comment:
Nit, but I'd like to see both calls either aliased or fully-qualified as appropriate.